File: csharp/demo/demo20a01-semaphore.cs
/*
 * SEMAPHORES
 * Version A: Paper sheets and packages
 */
using System;
using System.Threading;



class Demo20A01 : IRunnable
{
    public void run()
    {
        var semPackage = new Semaphore(0, 8); // initialCount = 0, maximumCount = 8


        ThreadStart makeOneSheet = () =>
        {
            for (int i = 0; i < 4; ++i)
            {
                Console.WriteLine("Make 1 sheet");
                Thread.Sleep(1000);
                semPackage.Release();
            }
        };


        ThreadStart combineOnePackage = () =>
        {
            for (int i = 0; i < 4; ++i)
            {
                semPackage.WaitOne();
                semPackage.WaitOne();
                Console.WriteLine("Combine 2 sheets into 1 package");
            }
        };


        new Thread(makeOneSheet).Start();
        new Thread(makeOneSheet).Start();
        new Thread(combineOnePackage).Start();
    }
}

File: python/demo20a01_semaphore.py
'''
SEMAPHORES
Version A: Paper sheets and packages
'''

import time
import threading



sem_package = threading.Semaphore(0)



def make_one_sheet():
    for _ in range(4):
        print('Make 1 sheet')
        time.sleep(1)
        sem_package.release()



def combine_one_package():
    for _ in range(4):
        sem_package.acquire()
        sem_package.acquire()
        print('Combine 2 sheets into 1 package')



threading.Thread(target=make_one_sheet).start()
threading.Thread(target=make_one_sheet).start()
threading.Thread(target=combine_one_package).start()
